python/ray/worker.py

```diff
@@ -448,6 +426,30 @@ def retrieve_and_deserialize(self, object_ids, timeout, error_timeout=10):
                             driver_id=self.task_driver_id)
                     warning_sent = True
 
+    def _deserialize_object_from_arrow(self, data, metadata, object_id,
+                                       serialization_context):
+        if metadata:
+            # If metadata is not empty, return an exception object based on
+            # the error type.
+            error_type = int(metadata)
+            if error_type == ErrorType.WORKER_DIED:
+                return RayWorkerError()
+            elif error_type == ErrorType.ACTOR_DIED:
+                return RayActorError()
+            elif error_type == ErrorType.OBJECT_UNRECONSTRUCTABLE:
+                return UnreconstructableError(ray.ObjectID(object_id.binary()))
+            else:
+                assert False, "Unrecognized error type " + str(error_type)
+        elif data:
+            # If data is not empty, deserialize the object.
+            # Note, the lock is needed because `serialization_context` isn't
+            # thread-safe.
+            with self.plasma_client.lock:
+                return pyarrow.deserialize(data, serialization_context)
```

Comment:
Should we just make `serialization_context` a thread safe object? E.g., using the `thread_safe_client` wrapper?

Reply:
the `thread_safe_client` wrapper may not work for `serialization_context`, because the wrapper only wrap object's methods with the lock. However, if someone accesses a member field of `serialization_context`, it's still thread unsafe.